src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs

```diff
@@ -0,0 +1,44 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+
+namespace Internal.Runtime.InteropServices
+{
+    public static class InMemoryAssemblyLoader
+    {
+        public static unsafe int LoadAndExecuteInMemoryAssembly(IntPtr handle, int argc, [MarshalAs(UnmanagedType.LPArray, ArraySubType = UnmanagedType.LPWStr)] string[] argv)
+        {
+#if FEATURE_PAL
+            throw new PlatformNotSupportedException();
+#else
+            Assembly entryAssembly = AssemblyLoadContext.Default.LoadFromInMemoryModule(handle);
+            // Emulate traditional app start behavior that adds back on the path of the entry assembly
+            // to the array set for Environment.SetCommandLineArgs.
+            string[] environmentArgs = new string[argv.Length + 1];
+            environmentArgs[0] = entryAssembly.Location;
+            Array.Copy(argv, 0, environmentArgs, 1, argv.Length);
+
+            Environment.SetCommandLineArgs(environmentArgs);
```

Comment:
It would be much cleaner if we would refactor `CorHost2::ExecuteAssembly` into two parts and the inner section which handles command line arguments and calls execute assembly would be shared between the `CorHost2` method and the new method you're adding for this.
I understand that the end result is the same, but it would be nice to not duplicate code. (It would also be a bit slower due to managed->native->managed marshalling, but the impact should be very small)

Reply:
> be nice to not duplicate code

FWIW, I do not see much duplication here.